libsignal-service/src/service_address.rs

```diff
@@ -33,31 +36,93 @@ impl ServiceAddress {
     pub fn pni(&self) -> libsignal_protocol::Pni {
         libsignal_protocol::Pni::from_uuid_bytes(self.uuid.into_bytes())
     }
+
+    pub fn to_service_id(&self) -> String {
+        match self.identity {
+            ServiceIdType::AccountIdentity => self.uuid.to_string(),
+            ServiceIdType::PhoneNumberIdentity => {
+                format!("PNI:{}", self.uuid)
+            },
+        }
+    }
 }
 
 impl From<Uuid> for ServiceAddress {
     fn from(uuid: Uuid) -> Self {
-        Self { uuid }
+        Self {
+            uuid,
+            identity: ServiceIdType::AccountIdentity,
+        }
+    }
+}
+
+impl TryFrom<&ProtocolAddress> for ServiceAddress {
+    type Error = ParseServiceAddressError;
+
+    fn try_from(addr: &ProtocolAddress) -> Result<Self, Self::Error> {
+        let value = addr.name();
+        if let Some(pni) = value.strip_prefix("PNI:") {
+            Ok(ServiceAddress {
+                uuid: Uuid::parse_str(pni)?,
+                identity: ServiceIdType::PhoneNumberIdentity,
+            })
+        } else {
+            match Uuid::parse_str(value) {
+                Ok(uuid) => Ok(ServiceAddress {
+                    uuid,
+                    identity: ServiceIdType::AccountIdentity,
+                }),
+                Err(e) => {
+                    tracing::error!("Unknown identity format: {}", value);
+                    Err(ParseServiceAddressError::InvalidUuid(e))
+                },
+            }
+        }
     }
 }
```

Comment:
If we can convert from a `ProtocolAddress`, then our `to_protocol_address` should probably also format a PNI as `PNI:{uuid}`? I'm not sure how this maps on Signal-Android, and how SA/SD/iS handle the ServiceAddress, but I'd vote we keep it as close as possible.

Reply:
My approach is based on Signal sending UUID's in the format of `Some("{uuid}")` and `Some("PNI:{uuid}")` so that's why I try to maintain that representation too. One assumption is that `ProtocolAddress` can contain the `PNI:` prefix too, but that's indeed unverified w.r.t. e.g. Signal Android code base.

I'll check the `to_protocol_address` method, sounds like I've missed that one.